File: frontend/features/battle/engine/checkWinner.ts
import type { BattleState, PlayerId } from "../types";

export function checkWinner(state: BattleState): PlayerId | null {
  if (state.player1.monster.hp <= 0) {
    return "player2";
  }

  if (state.player2.monster.hp <= 0) {
    return "player1";
  }

  return null;
}

File: backend/internal/battle/checkwinner.go
package battle

// CheckWinner はHPが尽きた陣営がいれば相手側のPlayerIDを返す。両者健在ならnil。
func CheckWinner(state BattleState) *PlayerID {
	if state.Player1.Monster.HP <= 0 {
		winner := Player2
		return &winner
	}

	if state.Player2.Monster.HP <= 0 {
		winner := Player1
		return &winner
	}

	return nil
}
